### ISSUE TITLE
TypeScript Support

### ISSUE THREAD

**Issue by SachaG** [VulcanJS/Vulcan#2512]
I'm opening this issue to talk about what TypeScript support in Vulcan would look like. 

I think it'd be especially great to have auto-complete in VS Code through TypeScript to make it easier to use the various Vulcan APIs without having to constantly refer to the docs or example code.

**Comment by Apollinaire** [comment]
That would be great indeed.

TLDR : vs code gets its typings from `jsconfig.json` and we should create a terminal command that looks at a meteor project and fills this file with the paths to the source code for each package.


For documentation, there is already a VS Code extension by @mattblackdev : 
- [source](https://github.com/mattblackdev/meteor-package-intellisense)
- [discussion](https://forums.meteor.com/t/announcing-vs-code-extension-intellisense-for-your-meteor-packages/45110)
In short, this extension adds a command to the command palette of vs code, and when executed it looks inside the folders of the workspace for a jsconfig.json and fills the `compilerOptions.path` value with `meteor/vulcan:package` and directs to the packages it finds within the folder.
It already works pretty well within Vulcan's source code (try writing `import {} from 'meteor/vulcan:core'` and cmd+space inside the curly braces to trigger autocompletion).

We should try to move the codebase to TS to get more accurate typing (thus autocompletion) of all the exports from the codebase. It might even help us find some bugs! Since it's all written in JS for now, VS code can only infer the types from what we give, and the JSDoc that is sometimes present:

![image](https://user-images.githubusercontent.com/5938985/73358890-e8220b80-429f-11ea-839f-803e342b6344.png)

**For now, this only works within Vulcan's main repo** because of the [jsconfig.json](https://github.com/VulcanJS/Vulcan/blob/devel/jsconfig.json) (it was generated with the VS code extension)

To make this available to every Vulcan project, we need to add or generate the same file. There are some difficulties with that, namely the one-repo install vs two-repo install : 
- for one-repo, the Vulcan package files are located at `~/.meteor/packages/vulcan_{package}/@{version}/web.browser` and there is a file `web.browser.json` that points to all the source files. We could know the exact package version by looking into `.meteor/versions` within the project, then get all the paths by looking inside `web.browser.json`.
- for two-repo, the Vulcan package files are located wherever METEOR_PACKAGE_DIRS points and also at the same location as for one-repo install when there is no match for the package name.

I'm not sure that the vs code extension is the right way to do this. It might be better to have an npm command/module do it, because then we could use the METEOR_PACKAGE_DIRS hint. Also, the vs code command has no idea of where it is ran, so it might have unexpected behaviours if the workspace has several Vulcan projects in it. 

Side note: I don't know what would happen if we were to provide a `tsconfig.json` though, that might be worth trying.

**Comment by Apollinaire** [comment]
Actually, there is also another way : we could provide the typings in an npm package. Maybe directly through https://github.com/DefinitelyTyped/DefinitelyTyped or just as an npm package we publish ourself. We would need to get a script that generates the typing file `something.d.ts` for Vulcan packages. The downside of this method regarding the other is that it's less generic and it would only cover Vulcan's packages, but maybe we can make things such that it's easy to generate such a file from any Meteor package.

**Comment by SachaG** [comment]
As pointed out by @doronrk there are two areas where TypeScript could help:

1. Add types to Vulcan core code
2. Add more validation to Vulcan app code

By 2. I mean things like generating types for your data from your GraphQL or SimpleSchema schemas, using them in your component code, etc. 

I think for now we can focus on 1. since this is what will have the most impact on beginners learning Vulcan, as well as the maintainability of the core code. 

Another easy path to start would be to add the definitions to the starter repo or let people copy paste them to their own repos. But yes I think it's more or less equivalent to publishing them as an npm package?

See also this issue that talks about using TS inside Meteor packages:

https://github.com/meteor/meteor/issues/10828

From Ben: 

> For core packages, meteor npm i @types/meteor can provide some of the types, but a much better solution (in a future release of Meteor) would be for Meteor packages that use typescript to generate node_modules/meteor/package-name.d.ts files automatically, so any TypeScript-aware tool could find the most accurate types, automatically.
